src/proxy.rs

```diff
@@ -76,43 +80,45 @@ impl Proxy {
 
         SESSION_IN_PROGRESS_COUNT.fetch_add(1, Ordering::Relaxed);
 
-        Box::new(f1.and_then(|(mut jet_stream, mut jet_sink)| {
-            // Shutdown stream and the sink so the f2 will finish as well (and the join future will finish)
-            let _ = jet_stream.shutdown();
-            let _ = jet_sink.shutdown();
-
-            Ok((jet_stream, jet_sink))
-        })
-            .join(f2.and_then(|(mut jet_stream, mut jet_sink)| {
-                // Shutdown stream and the sink so the f2 will finish as well (and the join future will finish)
-                let _ = jet_stream.shutdown();
-                let _ = jet_sink.shutdown();
-
-                Ok((jet_stream, jet_sink))
-            }))
-            .and_then(|((jet_stream_1, jet_sink_1), (jet_stream_2, jet_sink_2))| {
-                let server_addr = jet_stream_1
-                    .peer_addr()
-                    .map(|addr| addr.to_string())
-                    .unwrap_or_else(|| "unknown".to_string());
-                let client_addr = jet_stream_2
-                    .peer_addr()
-                    .map(|addr| addr.to_string())
-                    .unwrap_or_else(|| "unknown".to_string());
-                info!(
-                    "Proxy result : {} bytes read on {server} and {} bytes written on {client}. {} bytes read on {client} and {} bytes written on {server}",
-                    jet_stream_1.nb_bytes_read(),
-                    jet_sink_1.nb_bytes_written(),
-                    jet_stream_2.nb_bytes_read(),
-                    jet_sink_2.nb_bytes_written(),
-                    server = &server_addr,
-                    client = &client_addr
-                );
-
-                Ok(())
-            }).then(|result| {
-            SESSION_IN_PROGRESS_COUNT.fetch_sub(1, Ordering::Relaxed);
-            result
-        }))
+        Box::new(
+            f1.select2(f2)
```

Comment:
I remember that I have hesitated between a select and a join.  If we call a select, could it be possible that we don't forward some packets ?  Let say we have these events :
- Client sent a terminate packet to the server
- Client close his connection
- Server tries to send data to client --> it fails
Result : We stop everything... But we didn't forward the terminate packet from the client to the server. And it is often there that we have the reason of the disconnection...  

I understand that with a join, the first thing that was done is a shutdown on the stream so we could lost it as well...

Is it something that we should pay attention or it should be ok ?

Reply:
I also hesitated about this case, but actually there are no problems thanks to Forward future :)
this case will behave like this:
1. The client sends a terminate packet to Jet;
2. The client closes the connection;
3. Jet reads terminate packet via Stream Future and sends the bytes to the Sink Future. And then, Forward Future calls Sink Future until it returns AsyncSink::Ready (so, all terminate packet is redirected);
4. Stream Future is polled again, but underlying stream returns 0 bytes read, so we finish the Stream Future. Forward Future closes Sink Future on the Stream Future's Async::Ready(None).

Result: Jet forwards the terminate packet from the client to the server, and then finishes this Future. After that Select Future works and stops another one Forward Future (and this works exactly like join + shutdown).
So, it should be ok :)